src/hotspot/share/opto/inlinetypenode.cpp

```diff
@@ -631,6 +631,157 @@ void InlineTypeNode::store(GraphKit* kit, Node* base, Node* ptr, bool immutable_
+    // Fall-through to the int case
+  case T_BOOLEAN:
+  case T_CHAR:
+  case T_BYTE:
+  case T_SHORT:
+  case T_INT:
+    cmp = igvn->register_new_node_with_optimizer(new CmpINode(val1, val2));
+    break;
+  case T_DOUBLE:
+    val1 = igvn->register_new_node_with_optimizer(new MoveD2LNode(val1));
+    val2 = igvn->register_new_node_with_optimizer(new MoveD2LNode(val2));
+    // Fall-through to the long case
+  case T_LONG:
+    cmp = igvn->register_new_node_with_optimizer(new CmpLNode(val1, val2));
+    break;
+  default:
+    assert(is_reference_type(bt), "must be");
+    cmp = igvn->register_new_node_with_optimizer(new CmpPNode(val1, val2));
+  }
+  Node* bol = igvn->register_new_node_with_optimizer(new BoolNode(cmp, test));
+  IfNode* iff = igvn->register_new_node_with_optimizer(new IfNode(*ctrl, bol, PROB_MAX, COUNT_UNKNOWN))->as_If();
+  Node* if_f = igvn->register_new_node_with_optimizer(new IfFalseNode(iff));
+  Node* if_t = igvn->register_new_node_with_optimizer(new IfTrueNode(iff));
+
+  region->add_req(if_t);
+  if (phi != nullptr) {
+    phi->add_req(igvn->intcon(0));
+  }
+  *ctrl = if_f;
+}
+
+// Check if a substitutability check between 'this' and 'other' can be implemented in IR
+bool InlineTypeNode::can_emit_substitutability_check(Node* other) {
+  if (other != nullptr && other->is_InlineType() && bottom_type() != other->bottom_type()) {
+    // Different types, this is dead code because there's a check above that guarantees this.
+    return false;
+  }
+  for (uint i = 0; i < field_count(); i++) {
+    ciType* ft = field_type(i);
+    if (ft->is_inlinetype()) {
+      // Check recursively
+      if (!field_value(i)->as_InlineType()->can_emit_substitutability_check(nullptr)){
+        return false;
+      }
+    } else if (!ft->is_primitive_type() && ft->as_klass()->can_be_inline_klass()) {
+      // Comparing this field might require (another) substitutability check, bail out
+      return false;
+    }
+  }
+  return true;
+}
+
+// Emit IR to check substitutability between 'this' and the value object referred to by 'ptr' (might be a buffer address or an InlineTypeNode)
+void InlineTypeNode::check_substitutability(PhaseIterGVN* igvn, RegionNode* region, Node* phi, Node** ctrl, Node* mem, Node* base, Node* ptr, bool flat) {
+  BarrierSetC2* bs = BarrierSet::barrier_set()->barrier_set_c2();
+  // TODO 8228361 double-check
+  // DecoratorSet decorators = IN_HEAP | MO_UNORDERED;
+  DecoratorSet decorators = MO_UNORDERED | C2_READ_ACCESS | C2_CONTROL_DEPENDENT_LOAD | IN_HEAP | C2_UNKNOWN_CONTROL_LOAD;
+  MergeMemNode* local_mem = igvn->register_new_node_with_optimizer(MergeMemNode::make(mem))->as_MergeMem();
+
+  ciInlineKlass* vk = inline_klass();
+  for (uint i = 0; i < field_count(); i++) {
+    int field_off = field_offset(i);
+    if (flat) {
+      // Flat access, no header
+      field_off -= vk->payload_offset();
+    }
+    Node* val = field_value(i);
```

Comment:
I suggest to name it `field_value` to distinguish between `ptr`. Maybe we can also rename `ptr` to `other`, `other_ptr` or `other_value_object` to make it more clear.